apps/prairielearn/src/schemas/index.ts

```diff
@@ -1,28 +1,206 @@
-import * as fs from 'node:fs';
-import * as path from 'node:path';
+import { type JSONSchemaType } from 'ajv';
+import { type ZodTypeDef, type ZodType } from 'zod';
+import {
+  ignoreOverride,
+  type JsonSchema7Type,
+  type Options,
+  type Refs,
+  zodToJsonSchema,
+} from 'zod-to-json-schema';
+
+import {
+  QuestionSchema,
+  AssessmentSchema,
+  CourseInstanceSchema,
+  CourseSchema,
+  NewsItemSchema,
+  CalculationQuestionOptionsSchema,
+  CheckboxQuestionOptionsSchema,
+  FileQuestionOptionsSchema,
+  MultipleChoiceQuestionOptionsSchema,
+  MultipleTrueFalseQuestionOptionsSchema,
+  QuestionOptionsv3Schema,
+  ElementCoreSchema,
+  ElementCourseSchema,
+  ElementExtensionSchema,
+  type NewsItem,
+  type MultipleTrueFalseQuestionOptions,
+  type QuestionOptionsv3,
+  type MultipleChoiceQuestionOptions,
+  type FileQuestionOptions,
+  type CheckboxQuestionOptions,
+  type CalculationQuestionOptions,
+  type Question,
+  type ElementExtension,
+  type ElementCore,
+  type CourseInstance,
+  type Course,
+  type Assessment,
+  type ElementCourse,
+  ColorSchema,
+  PointsSingleSchema,
+  PointsSchema,
+  QuestionIdSchema,
+  ForceMaxPointsSchema,
+  AdvanceScorePercSchema,
+  AssessmentSetSchema,
+  WorkspaceOptionsSchema,
+  DependencySchema,
+} from './schemas/index.js';
+export * from './schemas/index.js';
 
 /**
- * Reads a JSON file from the schemas directory.
+ * Rewrite the group role annotation for canView and canSubmit fields.
+ * zod-to-json-schema doesn't support a concept of unique items in an array (only sets),
+ * so we need to override the schema.
```

Comment:
This seems super hacky to me. Can we just allow duplicates in the schema and dedupe (if necessary) elsewhere before we insert into the database? That might not even be necessary. There are at least two `string[]` fields that just allow duplicates and it's never caused us trouble (`tags` and `sharingSets` in question JSON).

Feel free to land that in a separate PR so that this PR isn't introducing any changes there.

Reply:
I will try to land that in a separate PR.

Reply:
Tracked by #11799